### DISCUSSION TITLE
Proposal: Improve Native Method Declaration with Source Generators

### DISCUSSION THREAD

**Discussion by Der-Joscha** [StudioCherno/Coral#10] [Ideas]
First of all, I wanted to thank you for creating this project.

I have a suggestion regarding the usage of `unsafe delegate*<...>` for declaring native methods. In my opinion, it seems counterintuitive and prone to errors. Specifically:

1. It's not immediately obvious which type parameters are arguments and which is the return type.
2. There's a risk of a careless developer accidentally overwriting the field.
3. Every time the delegate is called, it has to be wrapped in a `unsafe { ... }` block.

I propose using source generators to generate a thin wrapper method over those delegates. For instance, instead of:
```c#
 internal static unsafe delegate*<NativeArray<float>> ArrayReturnIcall;
```
Users of Coral.Managed could declare their methods like this:
```c# 
[NativeMethod] 
internal static partial NativeArray<float> ArrayReturnIcall();
```

The source generator would then kick in at compile time and generate the implementation of the wrapper as well as the delegate.

Here's an example:

```c# 
namespace Example.Managed
{
    public partial class ExampleClass
    {
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        internal static partial NativeArray<float> ArrayReturnIcall()
        {
           unsafe { return Delegates.ExampleClass.ArrayReturnIcall(); }
        }
        // other native methods...
    }
}


#pragma warning disable CS0649 // Field is never assigned to, and will always have its default value
namespace Example.Managed.Delegates
{
    [CompilerGenerated]
    public static class ExampleClass
    {
        [CompilerGenerated, DebuggerBrowsable(DebuggerBrowsableState.Never)]
        internal static unsafe delegate*<NativeArray<float>> ArrayReturnIcall;
        // other delegates...
    }
}
#pragma warning restore CS0649 // Field is never assigned to, and will always have its default value
```

Note that this approach is optional and works in combination with the current implementation. It still allows the use of raw `unsafe delegate*<...>`.

I'd love to hear your thoughts on this idea, and I'm more than willing to assist with a pull request if you find it promising.

Best regards

**Comment by peter1745** [comment]
Thanks for the suggestion! Funnily enough I'm working on something like this for calling Corals C# methods from C++, so I could potentially extend the source generator to do what you've suggested, and I probably *will* end up doing that, although it might not be for a little while 🙂

**Reply by Der-Joscha** [reply]
Thanks for your reply. 
Let me know if you want some help. I have already a basic implementation for that in a personal repository (except referencing the source generator with 'OutputItemType="Analyzer"' when generating the .csproj).
